app/models/spree/address_decorator.rb

```diff
@@ -1,35 +1,27 @@
 Spree::Address.class_eval do
-  has_one :enterprise
+  has_one :enterprise, dependent: :restrict
   belongs_to :country, class_name: "Spree::Country"
 
   after_save :touch_enterprise
 
   geocoded_by :geocode_address
 
-  delegate :name, :to => :state, :prefix => true, :allow_nil => true
+  delegate :name, to: :state, prefix: true, allow_nil: true
 
   def geocode_address
-    geocode_address = [address1, address2, zipcode, city, country.andand.name, state.andand.name]
-    filtered_address = geocode_address.select{ |field| !field.nil? && field != '' }
-    filtered_address.compact.join(', ')
+    render_address([address1, address2, zipcode, city, country.andand.name, state.andand.name])
```

Comment:
these methods should be in a separate service, some sort of geocoder_service that formats addresses according to different given inputs. And this service methods should be unit tested at least with simple positive examples.

another option, is to add functional tests to avoid this bug from re-appearing without us noticing.

Reply:
Agree on the first one but it doesn't exclude the second option.

Reply:
I agree that these methods would be better in a separate service. But I didn't want to go too far out of scope. All commits except the last one with the initializer are pure refactoring and don't contribute anything to fixing the original issue. It was just me understanding the code and cleaning it up along the way. I thought about not including them in the PR, but then I thought it might be useful to others to know where the geocoding is actually happening.

Before we pull the rendering out into a service, we should discuss using our own model and table for enterprise addresses. We have some ugly hacks to use `Spree::Address` for enterprise locations. For example, Spree::Address requires first and last name, but enterprises don't have that. So in the enterprise model we set these fields to `"unused"` to make the addresses valid.

We also added the `has_one :enterprise` relationship which adds unneeded bulk to all address objects that are used for orders. All addresses for orders get automatically geolocated even though we never use that. Point is: enterprise addresses have a very different purpose and hold different data to Spree's addresses.

By the way, the rendering method is unit tested:
https://github.com/openfoodfoundation/openfoodnetwork/blob/8fb81bb6a7da16f8ca0124b448e5315aa5c8fba2/spec/models/spree/addresses_spec.rb
But that wasn't the problem. The problem was that Google decided to give different responses to API requests. The only way to catch that is a live test. Usually all specs are set up to not use an internet connection. External services are mocked or HTTP responses are recorded and replayed. Hitting a live API is very bad practice in tests. The correct HTTP response depends on the API key, the setup in the developer console, the availability of a valid credit card and Google's terms of service. I'm not sure we can test that in a reasonable way.

Reply:
ahaha, I meant unit tests on the rendering methods only! the tests were there already. sorry about that.

the full refactor of address is a more complicated issue, agree. thanks for the details!

I think we could still just extract these methods to a geocoding service separate from the address class but fully appreciate what you said about refactoring. it's important we feel encouraged to refactor but don't feel pressed to make it perfect. a little improvement is always very welcome!